File: solutions/elixir/palindrome-products/1/lib/palindrome_products.ex
defmodule PalindromeProducts do
  @doc """
  Generates all palindrome products from an optionally given min factor (or 1) to a given max factor.
  """
  @spec generate(non_neg_integer, non_neg_integer) :: map
  def generate(max_factor, min_factor \\ 1)
  def generate(max_factor, min_factor) when min_factor > max_factor do
    raise ArgumentError
  end
  def generate(max_factor, min_factor) do
    palindromes(min_factor, max_factor)
    |> Enum.group_by(fn {product, _} -> product end, fn {_, factors} -> factors end)
  end

  defp palindromes(min, max) do
    for i <- min..max,
        j <- i..max,
        product = i * j,
        palindrome?(product),
        do: {product, [i, j]}
  end

  defp palindrome?(int) do
    digits = to_string(int)
    digits == String.reverse(digits)
  end
end

File: solutions/typescript/palindrome-products/1/palindrome-products.ts
interface Input {
  maxFactor: number
  minFactor?: number
}

interface Palindrome { 
  value: number | null; 
  factors: [number, number][];
}

interface Output {
  smallest: Palindrome;
  largest: Palindrome;
}

export function generate(params: Input): Output {
  const { minFactor = 1, maxFactor } = params;
  if (minFactor > maxFactor) {
    throw new Error('min must be <= max');
  }
  const palindromes: Palindrome[] = [];

  for (let i = minFactor; i <= maxFactor; i++) {
    for (let j = i; j <= maxFactor; j++) {
      const product = i * j;
      if (isPalindrome(product)) {
        let palindrome = palindromes.find(p => p.value === product);
        if (!palindrome) {
          palindrome = { value: product, factors: [] };
          palindromes.push(palindrome);
        }
        palindrome.factors.push([i, j]);
      }
    }
  }

  if (palindromes.length === 0) {
    return {
      smallest: {value: null, factors: []},
      largest: {value: null, factors: []}
    };
  }

  const validPalindromes = palindromes.filter(p => p.value !== null);
  const smallest = validPalindromes.reduce((min, p) => (p.value! < min.value!) ? p : min);
  const largest = validPalindromes.reduce((max, p) => (p.value! > max.value!) ? p : max);

  return {
    smallest,
    largest
  };
}

function isPalindrome(num: number): boolean {
  const str = num.toString();
  return str === str.split('').reverse().join('');
}
